apps/sync/js/adapters/history.js

```diff
@@ -0,0 +1,309 @@
+      return placesStore.get(id);
+    }).then(existedPlace => {
+      if (existedPlace) {
+        var newPlace = mergeRecordsToDataStore(existedPlace, place);
+        return placesStore.put(newPlace, id, revisionId);
+      } else {
+        return placesStore.add(place, id, revisionId);
+      }
+    }).catch(e => {
+      console.error(e);
+      return;
+    });
+  }
+
+  function updatePlaces(places) {
+    return new Promise(resolve => {
+      places.reduce((cur, next) => {
+        return cur.then(() => {
+          if (next.deleted) {
+            return deletePlaceByFxSyncId(next.fxsyncId);
+          } else {
+            return addPlace(next);
+          }
+        });
+      }, Promise.resolve()).then(resolve);
+    });
+  }
+
+  function traverseRecords(each, revisionId) {
+    return _ensureStore().then(placesStore => {
+      return new Promise(resolve => {
+        var cursor = placesStore.sync(revisionId);
+        runNextTask(cursor);
+
+        function runNextTask(cursor) {
+          cursor.next().then(function(task) {
+            manageTask(cursor, task);
+          });
+        }
+
+        function manageTask(cursor, task) {
+          if (task.operation === 'done') {
+            resolve(task);
+            return;
+          }
+          each(task).then(() => {
+            runNextTask(cursor);
+          });
+        }
+      });
+    });
+  }
+
+  function deletePlace(id) {
+    return _ensureStore().then(placesStore => {
+      return placesStore.remove(id);
+    });
+  }
+
+  function deletePlaceByFxSyncId(fxsyncId) {
+    var deletingTask;
+
+    function each(task) {
+      if (task.data && task.data.fxsyncId === fxsyncId) {
+        deletingTask = task;
+      }
+      return Promise.resolve();
+    }
+
+    return traverseRecords(each).then(() => {
+      if (deletingTask) {
+        return deletePlace(deletingTask.id);
+      }
+      return Promise.resolve();
+    });
+  }
+
+  return {
+    mergeRecordsToDataStore: mergeRecordsToDataStore,
+    setSyncedCollectionMtime: setSyncedCollectionMtime,
+    getSyncedCollectionMtime: getSyncedCollectionMtime,
+    updatePlaces: updatePlaces,
+    deletePlaceByFxSyncId: deletePlaceByFxSyncId
+  };
+})();
+
+DataAdapters.history = {
+/**
+    KintoCollection.list() provides a list contains all records sorted by
+  "last_modified" property which indicates the data-modified time in
+  descending order. The "last_modified" value can be recongized as which
+  records have already synchronized before. It is used to stop early and avoid
+  the redundant operations which had handled.
+
+    When the new adding/updating records are coming from History Collection in
+  the format [2], it should be converted to the format [1] then stored into PDS.
+  Besides, the ID from HC records should be written into PDS records as well
+  for deleting record usage.
+
+    When the new deleting records are coming, the ID can help us to find which
+  record has to be removed.
```

Comment:
"History records are stored locally in a DataStore with format [1] while history records coming from Firefox Sync (via Kinto collection) have format [2]. We need to convert from [1] to [2] and viceversa. Also, we need to add the record ID from the Kinto collection to its corresponding match in the local DataStore, so we can remove the local record afterwards if it is remotely deleted."